main.go

```diff
@@ -111,15 +113,48 @@ func run(cfg langserver.Config) error {
 		}
 		defer lis.Close()
 
-		log.Println("langserver-go: listening on", *addr)
+		log.Println("langserver-go: listening for TCP connections on", *addr)
+
+		connectionCount := 0
+
 		for {
 			conn, err := lis.Accept()
 			if err != nil {
 				return err
 			}
-			jsonrpc2.NewConn(context.Background(), jsonrpc2.NewBufferedStream(conn, jsonrpc2.VSCodeObjectCodec{}), handler, connOpt...)
+			connectionCount = connectionCount + 1
+			connectionID := connectionCount
+			log.Printf("langserver-go: received incoming connection #%d\n", connectionID)
+			jsonrpc2Connection := jsonrpc2.NewConn(context.Background(), jsonrpc2.NewBufferedStream(conn, jsonrpc2.VSCodeObjectCodec{}), handler, connOpt...)
+			go func() {
+				<-jsonrpc2Connection.DisconnectNotify()
+				log.Printf("langserver-go: disconnected connection #%d\n", connectionID)
+			}()
 		}
 
+	case "websocket":
+		mux := http.NewServeMux()
+
+		connectionCount := 0
+
+		mux.HandleFunc("/", func(responseWriter http.ResponseWriter, request *http.Request) {
+			var upgrader = websocket.Upgrader{CheckOrigin: func(r *http.Request) bool { return true }}
+			connection, err := upgrader.Upgrade(responseWriter, request, nil)
+			if err != nil {
+				log.Println("error upgrading HTTP to WebSocket:", err)
+			}
+			defer connection.Close()
+			connectionCount = connectionCount + 1
+			connectionID := connectionCount
+			log.Printf("langserver-go: received incoming connection #%d\n", connectionID)
+			<-jsonrpc2.NewConn(context.Background(), wsjsonrpc2.NewObjectStream(connection), langserver.NewHandler(cfg), connOpt...).DisconnectNotify()
```

Comment:
In contrast to the TCP case, this creates a new handler `langserver.NewHandler(cfg)` for every connection. Without this, the langserver throws an error:

```
language server is already initialized
```

@keegancsmith Do you happen to know if all usages of TCP mode only ever send `initialize` once? I'm guessing lsp-proxy is the only user of TCP mode, and performs connection sharing so that it only sends `initialize` once. Do you foresee any problems with creating a separate handler for each connection? I haven't run across any problems in my experimentation so far, but I just thought I'd check.

Reply:
Got it, fixed by creating a new handler for each connection.